pkg/operator/controllers/workaround/systemreserved.go

```diff
@@ -0,0 +1,252 @@
+		"Standard_D32s_v3":  {"MemGB": 128, "vCPUs": 32},
+		"Standard_D4as_v4":  {"MemGB": 16, "vCPUs": 4},
+		"Standard_D4s_v3":   {"MemGB": 16, "vCPUs": 4},
+		"Standard_D8as_v4":  {"MemGB": 32, "vCPUs": 8},
+		"Standard_D8s_v3":   {"MemGB": 32, "vCPUs": 8},
+		"Standard_E16s_v3":  {"MemGB": 128, "vCPUs": 16},
+		"Standard_E32s_v3":  {"MemGB": 256, "vCPUs": 32},
+		"Standard_E4s_v3":   {"MemGB": 32, "vCPUs": 4},
+		"Standard_E8s_v3":   {"MemGB": 64, "vCPUs": 8},
+		"Standard_F16s_v2":  {"MemGB": 32, "vCPUs": 16},
+		"Standard_F32s_v2":  {"MemGB": 64, "vCPUs": 32},
+		"Standard_F4s_v2":   {"MemGB": 8, "vCPUs": 4},
+		"Standard_F8s_v2":   {"MemGB": 16, "vCPUs": 8},
+	}
+	// convert total memory on the VM to amount to reserve
+	memReserved = map[int]string{
+		8:   "1800Mi",
+		16:  "2600Mi",
+		32:  "3560Mi",
+		64:  "5480Mi",
+		128: "9320Mi",
+		256: "11880Mi",
+	}
+	// convert vCPUs to millicores to reserve
+	cpuReserved = map[int]string{
+		2:  "500m",
+		4:  "500m",
+		8:  "500m",
+		16: "500m",
+		32: "500m",
+		64: "750m",
+	}
+)
+
+func NewSystemReserved(log *logrus.Entry, mcocli mcoclient.Interface, clustercli clusterapi.Interface, dh dynamichelper.DynamicHelper) *systemreserved {
+	utilruntime.Must(mcv1.AddToScheme(scheme.Scheme))
+	verFixed, err := version.ParseVersion("4.99.0") // TODO set this correctly when known.
+	utilruntime.Must(err)
+
+	return &systemreserved{
+		mcocli:       mcocli,
+		clustercli:   clustercli,
+		dh:           dh,
+		log:          log,
+		versionFixed: verFixed,
+	}
+}
+
+func (sr *systemreserved) Name() string {
+	return "SystemReserved fix for bz-1857446"
+}
+
+func (sr *systemreserved) IsRequired(clusterVersion *version.Version) bool {
+	return (clusterVersion.Lt(sr.versionFixed))
+}
+
+func (sr *systemreserved) kubeletConfig(vmSize string) (*unstructured.Unstructured, error) {
+	cap, ok := vmCapacity[vmSize]
+	if !ok {
+		return nil, fmt.Errorf("vmSize %s not valid", vmSize)
+	}
+
+	kubeletConfig := map[string]map[string]string{
+		"systemReserved": {
+			"memory": memReserved[cap["MemGB"]],
+			"cpu":    cpuReserved[cap["vCPUs"]],
+		},
+	}
+	cfgJSON, err := json.Marshal(kubeletConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	kc := &mcv1.KubeletConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   kubeletConfigName,
+			Labels: map[string]string{labelName: labelValue},
+		},
+		Spec: mcv1.KubeletConfigSpec{
+			MachineConfigPoolSelector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{labelName: labelValue},
+			},
+			KubeletConfig: &runtime.RawExtension{Raw: cfgJSON},
+		},
+	}
+	un := &unstructured.Unstructured{}
+	err = scheme.Scheme.Convert(kc, un, nil)
+	if err != nil {
+		return nil, err
+	}
+	return un, nil
+}
+
+func (sr *systemreserved) vmSize() (string, error) {
+	machinesets, err := sr.clustercli.MachineV1beta1().MachineSets(workerMachineSetsNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return "", err
+	}
+	largestVMSize := ""
```

Comment:
Dont know how to feel about this. Thinking we need to discuss this for a bit.
Like if the customer has multiple `worker` machineSets and one of them is supersized one (128GB+), we apply same limits to all machines. Which does not sound right.
On the other hand, alternative is very complex where we can't even implement it here.

Im thinking between 2 options at the moment:
1. Leave it as it is.
2. Just set static values of `500m` and `2000Mi` for all sizes and hope for the best and wait for upstream sorts this out.

2nd one removed a lot of checking and boilerplating in this PR, but creates a chance that 2GB might not be enough on High end boxes. @asalkeld @jim-minter wdyt?

Reply:
how about:
1. all the same - do the calculation I have in the PR
2. different - do static 500m and 2000Mi
That will give the best solution for most people